server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/RecipientRewriteTableProcessor.java

```diff
@@ -0,0 +1,294 @@
+        return new Predicate<MailAddress>() {
+            @Override
+            public boolean apply(MailAddress mailAddress) {
+                return !context.isLocalServer(mailAddress.getDomain());
+            }
+        };
+    }
+
+    private void forwardToRemoteAddress(MailAddress sender, MailAddress recipient, MimeMessage message, ImmutableList<MailAddress> mailAddresses, final MailetContext context) throws MessagingException {
+        ImmutableList<MailAddress> remoteAddress = FluentIterable.from(mailAddresses)
+            .filter(isNotLocalServer(context))
+            .toList();
+
+        if (!remoteAddress.isEmpty()) {
+            StringBuilder logBuffer = new StringBuilder(128);
+            try {
+                context.sendMail(sender, remoteAddress, message);
+                logBuffer.append("Mail for ")
+                    .append(recipient).append(" forwarded to ")
+                    .append(remoteAddress);
+            } catch (MessagingException ex) {
+                logBuffer.append("Error forwarding mail to ")
+                    .append(remoteAddress)
+                    .append("attempting local delivery");
+            }
+            context.log(LogLevel.INFO, logBuffer.toString());
+
+        }
+    }
+
+    private String getDefaultDomain(DomainList domainList) throws MessagingException {
+        try {
+            return domainList.getDefaultDomain();
+        } catch (DomainListException e) {
+            throw new MessagingException("Unable to access DomainList", e);
+        }
+    }
+
+    /**
+     * Handle the given mappings to map the original recipient to the right one
+     *
+     * @param mappings
+     *            a collection of mappings for the given recipient
+     * @param sender
+     *            the sender of the mail
+     * @param recipient
+     *            the original recipient of the email
+     * @param message
+     *            the mail message
+     * @return a collection of mapped recpient addresses
+     *
+     * @throws MessagingException
+     */
+    public Collection<MailAddress> handleMappings(Mappings mappings, MailAddress sender, MailAddress recipient, MimeMessage message,
+                                                  MailetContext context, DomainList domainList) throws MessagingException {
+        ImmutableList<Mapping> addressMappingWithoutDomains = getAddressWithDefaultDomain(mappings, domainList);
+
+        ImmutableList<Mapping> newAddressMappings = convertToNewMappings(mappings, addressMappingWithoutDomains);
+
+        ImmutableList<MailAddress> mailAddresses = buildMailAddressFromMappingAddress(newAddressMappings);
+
+        forwardToRemoteAddress(sender, recipient, message, mailAddresses, context);
+
+        return getLocalAddresses(mailAddresses, context);
+    }
+
+    public ImmutableList<MailAddress> rewriteRecipients(Mail mail, Function<MailAddress, Collection<MailAddress>> processMail) {
```

Comment:
Too much public methods. What's really your interface for this "processor" ?

Also, put public methods, then private methods.